rfd/0038-resource-encryption.md

```diff
@@ -0,0 +1,76 @@
+
+Broadly, there are two options available:
+
+1. Sharing a single encryption key across all authentication servers.
+2. Taking a card from the AWS Encryption SDK playbook and using a two layer key system.
+
+Option 1 has multiple significant drawbacks which make it unsuitable for production use. The major issue is that all data is fundamentally tied to the only layer of security meaning that if a malicous party potentially acquired access hold to the key, all data would have to be reincrypted. This makes key rotation prohibitive. Another major issue is that of prolonged key reuse which contradicts industry best practices.
+
+Option 2 is doing something very similar to what the AWS Encryption SDK does. Ideally using that library would be ideal since it is a very nice encryption toolkit but unfortunately it isn't available for Go. The strategy involves using unique ephemeral data keys for each encrypted object (rotated when reincrypted/modified) that are encrypted and bundled with each encrypted object in a defined format. A central PEM encoded master keypair stored in a new resource is used to encrypt these per object keys.
+
+Option 2 here seems like a far more robust solution for production use and that is thus what we should employ.
+
+### Encryption format
+
+For facilitating key rotation the encrypted objects should be stored in a seperate file from it's data key bundle. This is due to modification restricts on objects imposed by services such as AWS S3.
+
+All encrypted objects should be suffixed by a `.enc` extension to communicate that they are encrypted. For simplicity no other data should be bundled in the same file.
+
+The file that stores the encrypted data key should be a file with the name `$objectName.keys` stored next to the encrypted object.
+
+This format allows a decrypting auth server to easily fetch and decrypt the data key using the master key. It can then fetch and decrypt the object itself.
+
+Key rotation in event of a security incident can be performed by fetching and rewriting every key file to reencrypt the data key.
+
+Both the data and key file should have a short 8 byte header declaring the version. The header is to be interpreted as a little-endian 64 bit signed integer and the version described in this RFD is represented as 1.
+
+The format detailed above should be used wherever it makes sense from a security standpoint. Devitations from how the data object is encrypted should be made wherever necessary but the key bundle format should never be deviated from. A notable deviation that will be made is for Parquet based files where the encryption will be handled by giving Parquet control of the symmetric data key. The reason for this is that flat-file encryption with Parquet is vulnerable to security attacks similar to CRIME.
```

Comment:
Just to check my understanding: Parquet will handle storing the internal object-encryption keys for us and is thus exempt from having sidecar key files for each internal parquet data frame, yes?

Will we still be creating an ephemeral key (+ sidecar key file) that will serve as a file-local KEK?

Reply:
Parquet will have the same sidecar car file yes. The difference is that we aren't doing standard file encryption with the data key but instead Parquet handles encryption internally within the format. We still need to store the encrypted data key that Parquet used somewhere. That said, if we want to use AWS Athena which I am not sure of yet we're going to need to special case this to use AWS KMS encryption since AWS Athena is unfortunately very limited.